platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/DefaultConfigurationCache.kt

```diff
@@ -190,22 +197,38 @@ class DefaultConfigurationCache internal constructor(
             }
             cacheEntryRequiresCommit = false
         } else if (cacheEntryRequiresCommit) {
-            val reusedProjects = mutableSetOf<Path>()
-            val updatedProjects = mutableSetOf<Path>()
-            intermediateModels.value.visitProjects(reusedProjects::add, updatedProjects::add)
-            projectMetadata.value.visitProjects(reusedProjects::add) { }
-            store.useForStore { layout ->
-                writeConfigurationCacheFingerprint(layout, reusedProjects)
-                cacheIO.writeCacheEntryDetailsTo(buildStateRegistry, intermediateModels.value.values, projectMetadata.value.values, layout.fileFor(StateType.Entry))
-            }
-            problems.projectStateStats(reusedProjects.size, updatedProjects.size)
+            val projectUsage = collectProjectUsage()
+            commitCacheEntry(projectUsage.reused)
+            problems.projectStateStats(projectUsage.reused.size, projectUsage.updated.size)
             cacheEntryRequiresCommit = false
             // Can reuse the cache entry for the rest of this build invocation
             cacheAction = ConfigurationCacheAction.LOAD
         }
         scopeRegistryListener.dispose()
     }
 
+    private
+    fun collectProjectUsage(): ProjectUsage {
+        val reusedProjects = mutableSetOf<Path>()
+        val updatedProjects = mutableSetOf<Path>()
+        intermediateModels.value.visitProjects(reusedProjects::add, updatedProjects::add)
+        projectMetadata.value.visitProjects(reusedProjects::add) { }
+        return ProjectUsage(reusedProjects, updatedProjects)
+    }
+
+    private
+    data class ProjectUsage(val reused: Set<Path>, val updated: Set<Path>)
+
+    private
+    fun commitCacheEntry(reusedProjects: Set<Path>) {
+        store.useForStore { layout ->
+            writeConfigurationCacheFingerprint(layout, reusedProjects)
+            val usedModels = intermediateModels.value.collectAccessedValues()
```

Comment:
💅 It makes sense to extract these `intermediateModels.value` and `projectMetadata.value` expressions to a single place. It could be in a separate PR though.

Reply:
Good idea! It's relatively rare to see `= lazy { ... }` compared to `by lazy { ... }`, but Kotlin does not seem to provide a clean mechanism to check initialization-ess of the `by lazy` property, like [it does for the `lateinit`s](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/is-initialized.html).

A follow-up PR would be better for this

Reply:
Here is the PR:
https://github.com/gradle/gradle/pull/28431